sql/src/main/java/io/crate/jobs/PageDownstreamContext.java

```diff
@@ -81,6 +82,7 @@ public PageDownstreamContext(Logger logger,
         this.streamers = streamers;
         this.ramAccountingContext = ramAccountingContext;
         this.numBuckets = numBuckets;
+        this.buckets = new HashSet<>(numBuckets);
```

Comment:
why is using a dedicated set different from using the keySet of the existing bucketsByIdx map? I don't fully understand your commit comment as iterating over the keySet never ensured any ordering.

Reply:
The bucketsByIdx are eventually cleared. Previously they could be recreated by iterating from 0 to num buckets. However, buckets are not sequential anymore. We need to retain the bucket ids. Retaining them inside the bucketsByIdx map has some drawbacks, as we need to null out buckets when they are not in use and iterate over the map to check which ones are not null to see if we should trigger the PageResultListener. So overall, it is better to maintain a set with all valid bucket ids.